### PR TITLE
[api] Fix input validation

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
Surprise!  Turns out some of the validation that was being done for creation of objects was not being checked during serialization.

I've gone ahead and just made a bunch of pre DB lookup checks that verify that the input types are valid before we go to the DB.  This should prevent some bad transactions, and some DB lookups.

### Test Plan
See goldenfiles

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4088)

### CODE REVIEW

**Comment by banool** [comment]
Looks great, but I feel like we should find a way to make this more bulletproof. For example, we add middleware that calls the verify function on all input values. AKA require that every input type impls this verify trait, even if the function just returns Ok for many types. Though I feel like even this isn't quite right. Can't we find a way to do the checks at deserialization time? This feels too easy to accidentally forget to do.

**Comment by gregnazario** [comment]
> Looks great, but I feel like we should find a way to make this more bulletproof. For example, we add middleware that calls the verify function on all input values. AKA require that every input type impls this verify trait, even if the function just returns Ok for many types. Though I feel like even this isn't quite right. Can't we find a way to do the checks at deserialization time? This feels too easy to accidentally forget to do.

We can do them at deserialization time, but currently the testnet has invalid data, so if I do at deserialize time, it will just break.  We have to accept what's in the chain, and then validate on inputs.

We can change to check at deserialize time if we wipe the data

**Comment by banool** [comment]
Okay sounds like a plan, let's start with this. Will take another pass tomorrow.

File: api/src/events.rs [hunk 1]
```diff
@@ -96,6 +97,12 @@ impl EventsApi {
         limit: Query<Option<u16>>,
     ) -> BasicResultWith404<Vec<VersionedEvent>> {
         // TODO: Assert that Event represents u64s as strings.
+        event_handle.0.verify(0).map_err(|err| {
```
Comment (banool) [inline api/src/events.rs:100]: Why only zero recursion on this one?

File: api/types/src/move_types.rs [hunk 2]
```diff
@@ -382,6 +383,20 @@ pub struct MoveStructTag {
     pub generic_type_params: Vec<MoveType>,
 }
 
+impl VerifyInputWithRecursion for MoveStructTag {
+    fn verify(&self, recursion_count: u8) -> anyhow::Result<()> {
+        verify_module_identifier(self.module.as_str())
+            .map_err(|_| anyhow::anyhow!("invalid struct tag: {}", self))?;
+        verify_identifier(self.name.as_str())
+            .map_err(|_| anyhow::anyhow!("invalid struct tag: {}", self))?;
```
Comment (banool) [inline api/types/src/move_types.rs:391]: Could you make the error more specific, like `Invalid struct tag for module name`, `Invalid struct tag for function name`, etc.

File: api/types/src/move_types.rs [hunk 3]
```diff
@@ -1039,8 +1097,8 @@ impl FromStr for EntryFunctionId {
 }
 
 #[inline]
-fn invalid_entry_function_id(s: &str) -> anyhow::Error {
-    format_err!("invalid entry function id {:?}", s)
+fn invalid_entry_function_id<S: Display + Sized>(s: S) -> anyhow::Error {
+    format_err!("invalid entry function id {}", s)
```
Comment (banool) [inline api/types/src/move_types.rs:1101]: Can we format this with humans reading this in mind, like `Invalid entry function ID`

**Review by banool — APPROVED** [review]
This is sweet I can see a few validators that would've made things much more obvious for me when making dapps.

File: api/src/events.rs (hunk 1 above)
Comment (gregnazario) [reply api/src/events.rs:100]: It starts at 0, but every recursive call internally, increments by 1

**Review by davidiw — COMMENTED** [review]
Need to think more about the code here, but wanted to make a counter proposal. Why don't we simulate and if it succeeds pass to mempool and if fail we return an appropriate error that instructs the user to debug via simulation?

**Comment by gregnazario** [comment]
> Need to think more about the code here, but wanted to make a counter proposal. Why don't we simulate and if it succeeds pass to mempool and if fail we return an appropriate error that instructs the user to debug via simulation?

This does bring a lot more overhead for transactions, and would increase the latency quite a bit for transactions.

Generally in API design, you want time to create new resources to be relatively bounded, so you can differentiate performance between deployment version by metrics.

Vs the Simulate transaction API, you know that it can be wildly different, so you can separate out the high cost associated with simulating the transaction.

Note that these checks are run on both simulate and submit to help users not submit transactions that would fail due to simple input errors.  It should decrease simple mistakes with low overhead cost

**Review by jjleng — APPROVED** [review]
As a follow-up, we should think of if we can add the input validation and error report to middleware. Also, would like to do some fuzzing tests for this.

**Pull request merged**